### PR TITLE
issue 11 in TableDB

### PR DESCRIPTION
**Pull request by liuluheng**
issue 11 in TableDB, the api is both test for tableDB and raft tabledb

@LiXizhi

### CODE REVIEW

File: script/ide/System/Database/Collection.lua [hunk 1]
```diff
@@ -61,6 +62,11 @@ function Collection:GetParent()
 	return self.parent;
 end
 
+function Collection:GetProvider()
```
Comment (LiXizhi) [inline script/ide/System/Database/Collection.lua]: Is this provider or just provider name? 

the function name is a bit confusing

File: script/ide/System/Database/StorageProvider.lua [hunk 2]
```diff
@@ -21,19 +21,25 @@ local StorageProvider = commonlib.gettable("System.Database.StorageProvider");
...
 	StorageProvider.DefaultStorage = storageProvider;
 end
 
+function StorageProvider:GetStorageClass(name)
+	return storages[name] or StorageProvider.DefaultStorage;
+end
 
 function StorageProvider:CreateStorage(collection)
-	local store_class = self:GetStorageClass(collection:GetName());
+	local store_class = self:GetStorageClass(collection:GetProvider() or collection:GetName());
```
Comment (LiXizhi) [inline script/ide/System/Database/StorageProvider.lua]: Maybe just collection:GetProviderName() is fine. remove collection:GetName().

File: script/ide/System/Database/StorageProvider.lua [hunk 3]
```diff
@@ -21,19 +21,25 @@ local StorageProvider = commonlib.gettable("System.Database.StorageProvider");
...
 	if(store_class) then
-		return store_class:new():init(collection);
+		return store_class:new():init(collection, init_args[collection:GetProvider()]);
```
Comment (LiXizhi) [inline script/ide/System/Database/StorageProvider.lua]: is it safe to call, `init_args[collection:GetProvider()]`  ,  may be refactor to  a function. and check nil

File: script/ide/System/Database/test/test_TableDatabase.lua [hunk 4]
```diff
@@ -641,4 +641,42 @@ function TestMultipleDB()
...
+		name = "tabledb", 
+		{
+			name = "providers", 
+			{ name = "provider", attr = { name = "raft", type = "TableDB.RaftSqliteStore", file = "(g1)npl_mod/TableDB/RaftSqliteStore.lua" }, "./, localhost, 9004, server4:" }
+		},
+		{
+			name = "tables",
+			{ name = "table", attr = { provider = "raft", name = "RaftUsers" } }, 
+			{ name = "table", attr = { provider = "raft", name = "RaftTemp" } },
+			{ name = "table", attr = { name = "DefaultToLocalProvider" } },
```
Comment (LiXizhi) [inline script/ide/System/Database/test/test_TableDatabase.lua]: maybe we need to rewrite the table name in the entire test file

File: script/ide/System/Database/TableDatabase.lua [hunk 5]
```diff
@@ -244,10 +244,60 @@ function TableDatabase:ToData()
...
+							local storage = commonlib.gettable(provider.attr.type);
+							StorageProvider:RegisterStorageClass(provider.attr.name, storage, init_args)
+
+						end
+					end
+				end
+				if item.name == "tables" then
+					for i,table in ipairs(item) do
+						if table.name == "table" then
+							self:collection(table.attr.name, table.attr.provider);
```
Comment (LiXizhi) [inline script/ide/System/Database/TableDatabase.lua]: BTW, I do not think it is necessary to pass provider here as the second parameter. 
Maybe you can just: collection.parent:FindProvider() on demand. 

And you are also missing `connect` logic when table is opened. 

And I think connect should be async.

File: script/ide/System/Database/test/test_TableDatabase.lua (hunk 4 above)
Comment (liuluheng) [inline script/ide/System/Database/test/test_TableDatabase.lua]: the latter is more easy to do

Comment (liuluheng) [inline script/ide/System/Database/test/test_TableDatabase.lua]: [make default configurable](https://github.com/NPLPackages/main/pull/42/commits/dcc19f22476c2a0e0272a82fa28bae5c9fe9e889) implement the latter

**Pull request merged**